File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head or not head.next: return head
        dummy = ListNode(-1)
        dummy.next = head
        a = c = dummy
        for i in range(left - 1):
            a = a.next
        for i in range(right):
            c = c.next
        b = a.next
        d = c.next
        prev, curr = b, b.next
        while curr != d:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        a.next = c
        b.next = d
        return dummy.next
        

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode a = dummy;
        //需要四个点
        for(int i = 0; i < left -1 ; i++) a = a.next;
        ListNode c = a;
        for(int i = 0; i < right - left + 1; i++) c = c.next;
        ListNode b = a.next, d = c.next;
        
        for(ListNode prev = b, curr = b.next; curr != d;) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        a.next = c;
        b.next = d;

        return dummy.next;
    }
    
}